### PR TITLE
Remove lodash methods

### PR DESCRIPTION
**Pull request by Esemesek**

### CODE REVIEW

File: src/beautifier.ts [hunk 1]
```diff
@@ -591,38 +567,49 @@ export class Unibeautify {
...
         }
-      });
-      return transformedOptions;
+      }, {});
     } else {
       return options;
     }
   }
 
+  private static zipObject(keys: string[], values: string[]) {
```
Comment (Glavin001) [inline src/beautifier.ts]: This could be moved into a helper/utility file as we add useful functions to replace libraries like lodash.

**Comment by szeck87** [comment]
Don't worry about the CodeClimate current 5 issues right now, that will need to be a separate PR.  But the unit tests need to pass, and this one isn't:  https://github.com/Esemesek/unibeautify/blob/master/test/beautifier/beautifier.spec.ts#L117-L209.

File: src/beautifier.ts [hunk 2]
```diff
@@ -591,38 +567,49 @@ export class Unibeautify {
...
+          const values = fields.map(field => options[field]);
+          const obj = this.zipObject(fields, values);
+          return {
+            [key]: fn(obj),
+            ...acc,
+          };
         } else {
           return new Error(
-            `Invalid option "${fieldKey}" with value ${JSON.stringify(op)}.`
+            `Invalid option "${key}" with value ${JSON.stringify(option)}.`
```
Comment (szeck87) [inline src/beautifier.ts]: A few tests will purposefully return an error.  When they do, `acc` is reset each time, thus most of the keys are not being returned.  So need to handle that.

File: src/beautifier.ts [hunk 3]
```diff
@@ -591,38 +567,49 @@ export class Unibeautify {
...
           );
         }
-      });
-      return transformedOptions;
+      }, {});
```
Comment (Glavin001) [inline src/beautifier.ts]: Add type:
```
{} as OptionValues
```

File: src/beautifier.ts [hunk 4]
```diff
@@ -591,38 +567,49 @@ export class Unibeautify {
...
-        } else if (typeof op === "boolean") {
-          if (op === true) {
-            transformedOptions[fieldKey] = options[fieldKey];
-          }
-        } else if (_.isArray(op)) {
-          const [fields, fn] = op as BeautifyOptionTransform;
-          const vals = _.map(fields, field => options[field]);
-          const obj = _.zipObject(fields, vals);
-          transformedOptions[fieldKey] = fn(obj);
+      return Object.keys(beautifierOptions).reduce((acc, key) => {
```
Comment (Glavin001) [inline src/beautifier.ts]: Add typings:

```
 return Object.keys(beautifierOptions).reduce((acc: OptionValues, key: string) => {
```

File: src/beautifier.ts [hunk 5]
```diff
@@ -591,38 +567,49 @@ export class Unibeautify {
...
+            ...acc,
+          };
+        } else if (option instanceof Array) {
+          const [fields, fn] = option;
+          const values = fields.map(field => options[field]);
+          const obj = this.zipObject(fields, values);
+          return {
+            [key]: fn(obj),
+            ...acc,
+          };
```
Comment (Glavin001) [inline src/beautifier.ts]: Wrong order.

Should be:

```
return {
            ...acc,
            [key]: fn(obj),
          };
```

File: src/beautifier.ts [hunk 6]
```diff
@@ -591,38 +567,49 @@ export class Unibeautify {
...
+        } else if (typeof option === "function") {
+          return {
+            [key]: option(options[key]),
+            ...acc,
+          };
+        } else if (option === true) {
+          return {
+            [key]: options[key],
+            ...acc,
+          };
```
Comment (Glavin001) [inline src/beautifier.ts]: Wrong order.

Should be:

```
return {
            ...acc,
            [key]: options[key],
          };
```

File: src/beautifier.ts [hunk 7]
```diff
@@ -591,38 +567,49 @@ export class Unibeautify {
...
+        if (typeof option === "string") {
+          return {
+            [key]: options[option],
+            ...acc,
+          };
+        } else if (typeof option === "function") {
+          return {
+            [key]: option(options[key]),
+            ...acc,
+          };
```
Comment (Glavin001) [inline src/beautifier.ts]: Wrong order.

Should be:

```
return {
            ...acc,
            [key]: option(options[key]),
          };
```

File: src/beautifier.ts [hunk 8]
```diff
@@ -591,38 +567,49 @@ export class Unibeautify {
...
-          const vals = _.map(fields, field => options[field]);
-          const obj = _.zipObject(fields, vals);
-          transformedOptions[fieldKey] = fn(obj);
+      return Object.keys(beautifierOptions).reduce((acc, key) => {
+        const option = beautifierOptions[key];
+        if (typeof option === "string") {
+          return {
+            [key]: options[option],
+            ...acc,
+          };
```
Comment (Glavin001) [inline src/beautifier.ts]: Wrong order.

Should be:

```
return {
            ...acc,
            [key]: options[option],
          };
```

File: src/beautifier.ts (hunk 2 above)
Comment (Glavin001) [reply src/beautifier.ts]: Looks like the error may not even have been returned in the previous code. Would have been absorbed by the `forEach` method.

**Comment by szeck87** [comment]
Looking good! The only lodash import is in update-languages.ts for `_.chain`, `_.keyBy`, and `_.union`. So at the very least lodash can be moved from regular dependency to a dev one.

File: src/utils.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+export function zipObject(keys: string[], values: string[]) {
+  return keys.reduce((acc, key, idx) => ({
+    [key]: values[idx],
+    ...acc,
+  }), {});
```
Comment (Glavin001) [inline src/utils.ts]: Pretty sure this is also the wrong order:

```
return keys.reduce((acc, key, idx) => ({
    ...acc,
    [key]: values[idx],
}), {});
```

File: src/utils.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+export function unique(array: any[]): any[] {
```
Comment (Glavin001) [inline src/utils.ts]: This looks like a good place for TypeScript's generics: https://www.typescriptlang.org/docs/handbook/generics.html

```
export function unique<T>(array: T[]): T[] {
```

**Comment by Glavin001** [comment]
Looks good to me! Thanks!

@szeck87 : I'll wait for your PR approval before merging?

**Comment by szeck87** [comment]
@Glavin001 is the [`update_languages.ts`](https://github.com/Unibeautify/unibeautify/blob/master/scripts/update-languages.ts) script run during the installation process at all?  If not could it be moved as a dev dependency instead of a regular one?  Think that's my only outstanding question then I'm ready to approve.

**Comment by szeck87** [comment]
@Esemesek can you move lodash to the devDependencies tree in package.json?  Then I'll approve.  Thanks!

**Comment by Esemesek** [comment]
Lodash is still used in `beautifier.ts`.

**Comment by szeck87** [comment]
Ah right, sorry I missed that.  And of course Object.assign doesn't do deep merge so it can't be replaced yet.  OK, approved.

**Comment by Glavin001** [comment]
Thank you @Esemesek for contributing! 🎉

**Comment by Glavin001** [comment]
Published to v0.12.0

**Pull request merged**